File: packages/pvm-assemblyscript/assembly/simple-ram.ts
/**
 * Simple RAM Implementation (AssemblyScript)
 *
 * A simplified RAM implementation with page access rights but without regions.
 * Uses a flat memory space with per-page access control (READ, WRITE, NONE).
 * Used for runBlob and test vectors where we want a simple memory model.
 */

import { FaultCheckResult, ReadResult, RAM, MemoryAccessType, WriteResult } from './types'
import { MEMORY_CONFIG } from './config'

/**
 * Simple RAM - flat memory space with page access rights
 * 
 * Uses a single contiguous Uint8Array for memory storage that grows as needed.
 * Tracks page access rights (READ, WRITE, NONE) per page using a Map for efficient
 * sparse storage of only initialized pages.
 */
export class SimpleRAM implements RAM {
  // Single contiguous memory storage (grows as needed)
  private memory: Uint8Array = new Uint8Array(0)
  
  // Page access rights: Map from page index to MemoryAccessType
  // Only stores pages that have been explicitly initialized
  // Uninitialized pages default to NONE
  private pageAccess: Map<u32, MemoryAccessType> = new Map<u32, MemoryAccessType>()
  
  // Track current heap pointer for SBRK operations
  currentHeapPointer: u32 = 0

  // JIP-6 trace support: Track last load/store for each instruction step
  lastLoadAddress: u32 = 0
  lastLoadValue: u64 = 0
  lastStoreAddress: u32 = 0
  lastStoreValue: u64 = 0

  /**
   * Clear last load/store tracking (call at start of each instruction)
   */
  clearLastMemoryOp(): void {
    this.lastLoadAddress = 0
    this.lastLoadValue = 0
    this.lastStoreAddress = 0
    this.lastStoreValue = 0
  }

  /**
   * Get page index for an address
   */
  private getPageIndex(address: u32): u32 {
    return address / MEMORY_CONFIG.PAGE_SIZE
  }


  /**
   * Ensure memory is large enough for the given address range
   * Grows the single contiguous memory array as needed
   */
  private ensureMemorySize(requiredSize: u32): void {
    if (u32(this.memory.length) >= requiredSize) {
      return // Already large enough
    }
    
    // Grow to next page boundary
    const alignedSize = ((requiredSize / MEMORY_CONFIG.PAGE_SIZE) + 1) * MEMORY_CONFIG.PAGE_SIZE
    const newMemory = new Uint8Array(alignedSize)
    
    // Copy existing data
    if (this.memory.length > 0) {
      newMemory.set(this.memory, 0)
    }
    
    // New memory is zero-initialized by default
    this.memory = newMemory
  }

  /**
   * Get page access type for a page index
   * Returns NONE by default for uninitialized pages
   */
  private getPageAccess(pageIndex: u32): MemoryAccessType {
    // Check if page exists in map, return NONE if not found
    return this.pageAccess.has(pageIndex) ? this.pageAccess.get(pageIndex) : MemoryAccessType.NONE
  }

  /**
   * Check if an address range has the required access type
   * Returns fault address (page start) if access is denied, 0 if allowed
   */
  private checkAccess(address: u32, size: u32, requiredAccess: MemoryAccessType): u32 {
    const startPage = this.getPageIndex(address)
    const endPage = this.getPageIndex(address + size - 1)
    
    for (let pageIndex = startPage; pageIndex <= endPage; pageIndex++) {
      const pageAccess = this.getPageAccess(pageIndex)
      
      // Check if page has required access
      if (requiredAccess === MemoryAccessType.READ) {
        // For READ, need READ or WRITE
        if (pageAccess === MemoryAccessType.NONE) {
          return pageIndex * MEMORY_CONFIG.PAGE_SIZE
        }
      } else if (requiredAccess === MemoryAccessType.WRITE) {
        // For WRITE, need WRITE
        if (pageAccess !== MemoryAccessType.WRITE) {
          return pageIndex * MEMORY_CONFIG.PAGE_SIZE
        }
      }
    }
    
    return 0 // No fault
  }

  /**
   * Read multiple octets from memory
   */
  readOctets(address: u32, count: u32): ReadResult {
    if (count === 0) {
      return new ReadResult(new Uint8Array(0), 0)
    }

    // Check read access
    const faultAddress = this.checkAccess(address, count, MemoryAccessType.READ)
    if (faultAddress !== 0) {
      return new ReadResult(null, faultAddress)
    }

    // Ensure memory is allocated
    this.ensureMemorySize(address + count)
    
    // Read directly from contiguous memory
    const result = new Uint8Array(count)
    const sourceView = this.memory.subarray(address, address + count)
    result.set(sourceView, 0)

    // JIP-6 trace support: Track last load address and value
    this.lastLoadAddress = address
    if (count > 0) {
      let value: u64 = 0
      const bytesToRead = min(count, 8)
      for (let i: u32 = 0; i < bytesToRead; i++) {
        value |= u64(result[i]) << (i * 8)
      }
      this.lastLoadValue = value
    }

    return new ReadResult(result, 0) // No faults
  }

  /**
   * Write multiple octets to memory
   */
  writeOctets(address: u32, values: Uint8Array): WriteResult {
    if (values.length === 0) {
      return new WriteResult(false, 0)
    }

    // Check write access (same pattern as ram.ts)
    const writableResult = this.isWritableWithFault(address, u32(values.length))
    if (!writableResult.success) {
      return new WriteResult(true, writableResult.faultAddress !== 0 ? writableResult.faultAddress : 0xFFFFFFFF)
    }

    // Ensure memory is allocated
    this.ensureMemorySize(address + values.length)
    
    // Write directly to contiguous memory
    this.memory.set(values, address)

    // Update heap pointer if writing beyond current heap
    if (address + values.length > this.currentHeapPointer) {
      this.currentHeapPointer = address + values.length
    }

    // JIP-6 trace support: Track last store address and value
    this.lastStoreAddress = address
    if (values.length > 0) {
      let value: u64 = 0
      const bytesToRead = min(u32(values.length), 8)
      for (let i: u32 = 0; i < bytesToRead; i++) {
        value |= u64(values[i]) << (i * 8)
      }
      this.lastStoreValue = value
    }

    return new WriteResult(false, 0) // No faults
  }

  /**
   * Allocate pages (for SBRK)
   */
  allocatePages(startPage: u32, count: u32): void {
    // Ensure memory is allocated for these pages
    const startAddress = startPage * MEMORY_CONFIG.PAGE_SIZE
    const endAddress = (startPage + count) * MEMORY_CONFIG.PAGE_SIZE
    this.ensureMemorySize(endAddress)
    
    // Update heap pointer
    if (endAddress > this.currentHeapPointer) {
      this.currentHeapPointer = endAddress
    }
  }

  /**
   * Check if memory is readable
   */
  isReadableWithFault(address: u32, size: u32 = u32(1)): FaultCheckResult {
    const faultAddress = this.checkAccess(address, size, MemoryAccessType.READ)
    return new FaultCheckResult(faultAddress === 0, faultAddress)
  }

  /**
   * Check if memory is writable
   * Matches ram.ts implementation: finds minimum inaccessible address, returns page start
   */
  isWritableWithFault(address: u32, size: u32 = u32(1)): FaultCheckResult {
    const endRequestedAddress = address + size
    let minInaccessibleAddress: u32 = 0xFFFFFFFF // Sentinel value for "not found"

    // Check each address in the range to find the first one that's not writable
    for (let addr: u32 = address; addr < endRequestedAddress; addr++) {
      const pageIndex = this.getPageIndex(addr)
      const pageAccess = this.getPageAccess(pageIndex)
      
      // For WRITE, need WRITE access
      if (pageAccess !== MemoryAccessType.WRITE) {
        minInaccessibleAddress = addr
        break
      }
    }

    if (minInaccessibleAddress !== 0xFFFFFFFF) {
      // Gray Paper: fault address is page start of the minimum inaccessible address
      const faultAddress = this.getPageIndex(minInaccessibleAddress) * MEMORY_CONFIG.PAGE_SIZE
      return new FaultCheckResult(false, faultAddress)
    }

    return new FaultCheckResult(true, 0)
  }

  /**
   * Initialize memory layout (no-op for SimpleRAM)
   */
  initializeMemoryLayout(
    argumentData: Uint8Array,
    readOnlyData: Uint8Array,
    readWriteData: Uint8Array,
    stackSize: u32,
    heapZeroPaddingSize: u32,
  ): void {
    // No-op: SimpleRAM doesn't use regions
    // If needed, we could write the data to appropriate addresses
    // For now, just update heap pointer
    if (readWriteData.length > 0) {
      // Assume heap starts at 2 * 65536 (standard heap start)
      const heapStart = 2 * 65536
      this.writeOctets(heapStart, readWriteData)
    }
  }

  /**
   * Set page access rights for an address range
   */
  setPageAccessRights(address: u32, length: u32, accessType: MemoryAccessType): void {
    if (length === 0) {
      return // Nothing to set
    }
    
    const startPage = this.getPageIndex(address)
    // Calculate end page: (address + length - 1) / PAGE_SIZE
    // Use safe calculation to avoid underflow
    const endAddress = address + length - 1
    const endPage = this.getPageIndex(endAddress)
    
    // Set access rights for all pages in the range
    for (let pageIndex = startPage; pageIndex <= endPage; pageIndex++) {
      this.pageAccess.set(pageIndex, accessType)
    }
  }

  /**
   * Initialize a memory page (creates the page if needed and sets access rights)
   * This is the interface method required by RAM interface
   */
  initPage(address: u32, length: u32, accessType: MemoryAccessType): void {
    if (length === 0) {
      return // Nothing to initialize
    }
    
    // Ensure memory is allocated for this address range
    this.ensureMemorySize(address + length)
    
    const startPage = this.getPageIndex(address)
    // Calculate end page: (address + length - 1) / PAGE_SIZE
    // Use safe calculation to avoid underflow
    const endAddress = address + length - 1
    const endPage = this.getPageIndex(endAddress)
    
    // Set access rights for all pages in the range
    for (let pageIndex = startPage; pageIndex <= endPage; pageIndex++) {
      this.pageAccess.set(pageIndex, accessType)
    }
  }

  /**
   * Write to memory during initialization (bypasses access rights)
   */
  writeOctetsDuringInitialization(address: u32, values: Uint8Array): void {
    // Bypass access rights checking during initialization
    // This allows writing to read-only pages during setup
    if (values.length === 0) {
      return
    }

    // Ensure memory is allocated (bypasses access check)
    this.ensureMemorySize(address + values.length)
    
    // Write directly to contiguous memory
    this.memory.set(values, address)

    // Update heap pointer if writing beyond current heap
    if (address + values.length > this.currentHeapPointer) {
      this.currentHeapPointer = address + values.length
    }
  }

  /**
   * Get page dump for a specific page index
   * Returns a copy of the page data (4KB) or zeros if page doesn't exist
   */
  getPageDump(pageIndex: u32): Uint8Array {
    const pageSize = MEMORY_CONFIG.PAGE_SIZE
    const startAddress = pageIndex * pageSize
    const endAddress = startAddress + pageSize
    
    // Ensure memory is large enough
    this.ensureMemorySize(endAddress)
    
    // Extract page data
    const pageData = new Uint8Array(pageSize)
    const sourceView = this.memory.subarray(startAddress, endAddress)
    pageData.set(sourceView, 0)
    
    return pageData
  }

  /**
   * Reset RAM to initial state
   * Clears all memory, page access rights, and resets heap pointer
   */
  reset(): void {
    // Clear memory (reset to empty array)
    this.memory = new Uint8Array(0)
    
    // Clear page access rights
    this.pageAccess.clear()
    
    // Reset heap pointer
    this.currentHeapPointer = 0
  }
}

File: packages/pvm-rust/src/simple_ram.rs
//! Simple flat RAM (mirrors assembly/simple-ram.ts).
//! Flat memory with per-page access rights for runBlob and test vectors.

use crate::config;
use crate::types::{FaultCheckResult, MemoryAccessType, Ram, ReadResult, WriteResult};
use std::collections::HashMap;

/// Simple RAM: single contiguous memory with page access rights.
#[derive(Default)]
pub struct SimpleRam {
    memory: Vec<u8>,
    page_access: HashMap<u32, MemoryAccessType>,
    current_heap_pointer: u32,
    last_load_address: u32,
    last_load_value: u64,
    last_store_address: u32,
    last_store_value: u64,
}

impl SimpleRam {
    #[must_use]
    pub fn new() -> Self {
        Self::default()
    }

    fn get_page_index(&self, address: u32) -> u32 {
        address / config::PAGE_SIZE
    }

    fn ensure_memory_size(&mut self, required_size: u32) {
        let len = self.memory.len() as u32;
        if len >= required_size {
            return;
        }
        let aligned = ((required_size / config::PAGE_SIZE) + 1) * config::PAGE_SIZE;
        self.memory.resize(aligned as usize, 0);
    }

    fn get_page_access(&self, page_index: u32) -> MemoryAccessType {
        *self.page_access.get(&page_index).unwrap_or(&MemoryAccessType::None)
    }

    fn check_access(&self, address: u32, size: u32, required: MemoryAccessType) -> u32 {
        let start_page = self.get_page_index(address);
        let end_addr = address.saturating_add(size).saturating_sub(1);
        let end_page = self.get_page_index(end_addr);
        for page_index in start_page..=end_page {
            let access = self.get_page_access(page_index);
            match required {
                MemoryAccessType::Read => {
                    if access == MemoryAccessType::None {
                        return page_index * config::PAGE_SIZE;
                    }
                }
                MemoryAccessType::Write => {
                    if access != MemoryAccessType::Write {
                        return page_index * config::PAGE_SIZE;
                    }
                }
                MemoryAccessType::None => {}
            }
        }
        0
    }
}

impl Ram for SimpleRam {
    fn read_octets(&mut self, address: u32, count: u32) -> ReadResult {
        if count == 0 {
            return ReadResult::new(Some(vec![]), 0);
        }
        let fault = self.check_access(address, count, MemoryAccessType::Read);
        if fault != 0 {
            return ReadResult::new(None, fault);
        }
        self.ensure_memory_size(address + count);
        let end = (address as usize + count as usize).min(self.memory.len());
        let data = self.memory[address as usize..end].to_vec();
        self.last_load_address = address;
        self.last_load_value = data
            .iter()
            .take(8)
            .enumerate()
            .fold(0u64, |acc, (i, &b)| acc | (u64::from(b) << (i * 8)));
        ReadResult::new(Some(data), 0)
    }

    fn write_octets(&mut self, address: u32, values: &[u8]) -> WriteResult {
        if values.is_empty() {
            return WriteResult::new(false, 0);
        }
        let size = values.len() as u32;
        let writable = self.is_writable_with_fault(address, size);
        if !writable.success {
            return WriteResult::new(
                true,
                if writable.fault_address != 0 {
                    writable.fault_address
                } else {
                    0xFFFF_FFFF
                },
            );
        }
        self.ensure_memory_size(address + size);
        let start = address as usize;
        let end = start + values.len();
        if end <= self.memory.len() {
            self.memory[start..end].copy_from_slice(values);
        }
        if address + size > self.current_heap_pointer {
            self.current_heap_pointer = address + size;
        }
        self.last_store_address = address;
        self.last_store_value = values
            .iter()
            .take(8)
            .enumerate()
            .fold(0u64, |acc, (i, &b)| acc | (u64::from(b) << (i * 8)));
        WriteResult::new(false, 0)
    }

    fn current_heap_pointer(&self) -> u32 {
        self.current_heap_pointer
    }

    fn set_current_heap_pointer(&mut self, value: u32) {
        self.current_heap_pointer = value;
    }

    fn allocate_pages(&mut self, start_page: u32, count: u32) {
        let _start_addr = start_page * config::PAGE_SIZE;
        let end_addr = (start_page + count) * config::PAGE_SIZE;
        self.ensure_memory_size(end_addr);
        if end_addr > self.current_heap_pointer {
            self.current_heap_pointer = end_addr;
        }
    }

    fn is_readable_with_fault(&self, address: u32, size: u32) -> FaultCheckResult {
        let fault = self.check_access(address, size, MemoryAccessType::Read);
        FaultCheckResult::new(fault == 0, fault)
    }

    fn initialize_memory_layout(
        &mut self,
        _argument_data: &[u8],
        _read_only_data: &[u8],
        read_write_data: &[u8],
        _stack_size: u32,
        _heap_zero_padding_size: u32,
    ) {
        if !read_write_data.is_empty() {
            let heap_start = 2 * 65_536u32;
            self.write_octets(heap_start, read_write_data);
        }
    }

    fn is_writable_with_fault(&self, address: u32, size: u32) -> FaultCheckResult {
        let end = address.saturating_add(size);
        let mut min_inaccessible = 0xFFFF_FFFFu32;
        for addr in address..end {
            let page_index = self.get_page_index(addr);
            if self.get_page_access(page_index) != MemoryAccessType::Write {
                min_inaccessible = addr;
                break;
            }
        }
        if min_inaccessible != 0xFFFF_FFFF {
            let fault = self.get_page_index(min_inaccessible) * config::PAGE_SIZE;
            return FaultCheckResult::new(false, fault);
        }
        FaultCheckResult::new(true, 0)
    }

    fn set_page_access_rights(&mut self, address: u32, length: u32, access_type: MemoryAccessType) {
        if length == 0 {
            return;
        }
        let start_page = self.get_page_index(address);
        let end_page = self.get_page_index(address + length - 1);
        for page_index in start_page..=end_page {
            self.page_access.insert(page_index, access_type);
        }
    }

    fn init_page(&mut self, address: u32, length: u32, access_type: MemoryAccessType) {
        if length == 0 {
            return;
        }
        self.ensure_memory_size(address + length);
        let start_page = self.get_page_index(address);
        let end_page = self.get_page_index(address + length - 1);
        for page_index in start_page..=end_page {
            self.page_access.insert(page_index, access_type);
        }
    }

    fn write_octets_during_initialization(&mut self, address: u32, values: &[u8]) {
        if values.is_empty() {
            return;
        }
        self.ensure_memory_size(address + values.len() as u32);
        let start = address as usize;
        let end = start + values.len();
        if end <= self.memory.len() {
            self.memory[start..end].copy_from_slice(values);
        }
        if address + values.len() as u32 > self.current_heap_pointer {
            self.current_heap_pointer = address + values.len() as u32;
        }
    }

    fn get_page_dump(&self, page_index: u32) -> Vec<u8> {
        let mut page = vec![0u8; config::PAGE_SIZE as usize];
        let start = (page_index * config::PAGE_SIZE) as usize;
        let _end = start + config::PAGE_SIZE as usize;
        if start < self.memory.len() {
            let copy_len = (self.memory.len() - start).min(config::PAGE_SIZE as usize);
            page[..copy_len].copy_from_slice(&self.memory[start..start + copy_len]);
        }
        page
    }

    fn reset(&mut self) {
        self.memory.clear();
        self.page_access.clear();
        self.current_heap_pointer = 0;
        self.last_load_address = 0;
        self.last_load_value = 0;
        self.last_store_address = 0;
        self.last_store_value = 0;
    }

    fn last_load_address(&self) -> u32 {
        self.last_load_address
    }

    fn last_load_value(&self) -> u64 {
        self.last_load_value
    }

    fn last_store_address(&self) -> u32 {
        self.last_store_address
    }

    fn last_store_value(&self) -> u64 {
        self.last_store_value
    }

    fn clear_last_memory_op(&mut self) {
        self.last_load_address = 0;
        self.last_load_value = 0;
        self.last_store_address = 0;
        self.last_store_value = 0;
    }
}
